This is for NoticeTask for issue #34459

mike2ykme opened the pull request:
Replace groovy build-tools (buildSrc) with java #34459

mike2ykme committed:
Commit to add NoticeTask to Java source

mike2ykme committed:
Commit to add NoticeTask to Java source

atorok commented on buildSrc/src/main/java/org/elasticsearch/gradle/NoticeTask.java:
```diff
@@ -0,0 +1,200 @@
+        return inputFile;
+    }
+
+    public void setInputFile(File inputFile) {
```
We don't configure this currently, so rather than having a filed, it would be preferable to have just a geter   annotated with `@InputFile`

mike2ykme replied:
The annotations have been moved to the getters

atorok commented on buildSrc/src/main/java/org/elasticsearch/gradle/NoticeTask.java:
```diff
@@ -0,0 +1,200 @@
+        return outputFile;
+    }
+
+    public void setOutputFile(File outputFile) {
```
Same as for input file.

mike2ykme replied:
Annotation moved

atorok commented on buildSrc/src/main/java/org/elasticsearch/gradle/NoticeTask.java:
```diff
@@ -0,0 +1,200 @@
+        licensesDirs.add(licensesDir);
+    }
+
+    public void addLicenseDir(File licenseDir){
```
We don't need another method that does the exact thing.

atorok commented on buildSrc/src/main/java/org/elasticsearch/gradle/NoticeTask.java:
```diff
@@ -0,0 +1,200 @@
+        return Collections.unmodifiableList(this.licensesDirs);
+    }
+
+//    public static String getText(File file){
```
please remove commented code.

atorok commented on buildSrc/src/main/java/org/elasticsearch/gradle/NoticeTask.java:
```diff
@@ -0,0 +1,200 @@
+    public void addLicenseDir(File licenseDir){
+        this.licensesDir(licenseDir);
+    }
+    public List<File> getLicensesDirs(){
```
This is not annotated as such in the Groovy code, but it should be `@InputFiles`

atorok commented on buildSrc/src/main/java/org/elasticsearch/gradle/NoticeTask.java:
```diff
@@ -0,0 +1,200 @@
+        final StringBuilder output = new StringBuilder();
+
+        try{
+            Files.lines(this.inputFile.toPath(), StandardCharsets.UTF_8)
```
One needs to close the stream when using [Files.lines](https://docs.oracle.com/javase/8/docs/api/java/nio/file/Files.html#lines-java.nio.file.Path-)

mike2ykme replied:
Thank you! I was unaware of the stream not closing this, I am very glad you pointed it out.

atorok commented on buildSrc/src/main/java/org/elasticsearch/gradle/NoticeTask.java:
```diff
@@ -0,0 +1,200 @@
+            output.append("\n\n");
+
+        } catch (IOException e) {
+            e.printStackTrace();
```
This is a logging anti pattern as it will most likely cause the same stack trace to be printed twice. 
I would just declare that the method throws IOException

atorok commented on buildSrc/src/main/java/org/elasticsearch/gradle/NoticeTask.java:
```diff
@@ -0,0 +1,200 @@
+        licensesDirs.stream()
+            .flatMap(file -> {
+                try {
+                    return Files.walk(file.toPath());
```
This one needs to be closed too

atorok commented on buildSrc/src/main/java/org/elasticsearch/gradle/NoticeTask.java:
```diff
@@ -0,0 +1,200 @@
+                } catch (IOException e) {
+                    e.printStackTrace();
+//                    throw new RuntimeException("failed to walk for " + file.toString());
+                    return Stream.empty();
```
We should never silence IOExceptions like this.

atorok commented on buildSrc/src/main/java/org/elasticsearch/gradle/NoticeTask.java:
```diff
@@ -0,0 +1,200 @@
+        writeToFile(outputFile,output.toString(),StandardCharsets.UTF_8);
+    }
+
+    public static void appendFileToOutput(File file, final String name, final String type, StringBuilder output) {
```
This and other methods used in this class only should be private.

atorok commented on buildSrc/src/main/java/org/elasticsearch/gradle/NoticeTask.java:
```diff
@@ -0,0 +1,200 @@
+
+    public static String getText(File file){
+        final StringBuilder builder = new StringBuilder();
+        try(BufferedReader reader = new BufferedReader(new FileReader(file))){
```
would prefer to use`Files` for consistency here.

atorok commented on buildSrc/src/main/java/org/elasticsearch/gradle/NoticeTask.java:
```diff
@@ -0,0 +1,200 @@
+
+    public static void writeToFile(File file, String output, Charset charset){
+
+        try(BufferedWriter out = Files.newBufferedWriter(file.toPath(),charset)){
```
would prefer to use`Files.write` for consistency here.

atorok commented on buildSrc/src/test/java/org/elasticsearch/gradle/NoticeTaskTest.java:
```diff
@@ -0,0 +1,165 @@
+public class NoticeTaskTest extends GradleUnitTestCase {
+
+    @Rule
+    public TemporaryFolder temporaryFolder = new TemporaryFolder();
```
The Gradle test project creates it's own temporary folder no need to pass it in.

atorok commented on buildSrc/src/test/java/org/elasticsearch/gradle/NoticeTaskTest.java:
```diff
@@ -0,0 +1,165 @@
+
+    public List<File> getListWithoutCopies() throws IOException{
+        List<File> files = this.getListWithCopies();
+        files.remove(2);
```
This makes the test hard to read, I would rather create the list without copies here and add to it above.

atorok commented on buildSrc/src/test/java/org/elasticsearch/gradle/NoticeTaskTest.java:
```diff
@@ -0,0 +1,165 @@
+
+
+        try {
+            noticeTask.generateNotice();
```
This won't fail if an exception is not thrown as it should. 
Use:
```
@Rule
    public ExpectedException expectedEx = ExpectedException.none();
```
instead.

mike2ykme replied:
I updated to now expect a RuntimeException with a certain a certain error message.

atorok replied:
I think you should use the `ExpectedException` rule. 
I don't think this test passes as it's written now. 
The correct way would be to have a `fail()` as the last line in `try`, but the rule reads better.

atorok commented on buildSrc/src/test/java/org/elasticsearch/gradle/NoticeTaskTest.java:
```diff
@@ -0,0 +1,165 @@
+    }
+
+    private NoticeTask createTask(Project project) {
+        return (NoticeTask) project.task(singletonMap("type", NoticeTask.class), "NoticeTask");
```
The cast can be avoided by using `project.getTasks().create`

alpar-t reviewed (commented):
Thanks for your contribution @mike2ykme !

I left a few comments.

mike2ykme committed:
Updated per notes on https://github.com/elastic/elasticsearch/pull/37032/files

mike2ykme committed:
Updated per notes on https://github.com/elastic/elasticsearch/pull/37032/files

mike2ykme committed:
Updated per: https://github.com/elastic/elasticsearch/pull/37032/files

removed commented out lines

atorok commented on buildSrc/src/main/java/org/elasticsearch/gradle/NoticeTask.java:
```diff
@@ -0,0 +1,173 @@
+        licensesDirs.add(licensesDir);
+    }
+
+    public List<File> getLicensesDirs() {
```
This needs to be annotated with `@InputFiles` to have correct up-to-date checks.

mike2ykme replied:
Done

atorok commented on buildSrc/src/main/java/org/elasticsearch/gradle/NoticeTask.java:
```diff
@@ -0,0 +1,173 @@
+        final StringBuilder output = new StringBuilder();
+
+
+        try (Stream<String> lines = Files.lines(this.inputFile.toPath(), StandardCharsets.UTF_8)) {
```
please use a single line for spacing

atorok commented on buildSrc/src/main/java/org/elasticsearch/gradle/NoticeTask.java:
```diff
@@ -0,0 +1,173 @@
+        licensesDirs.forEach(file -> {
+            try (Stream<Path> pathStream = Files.walk(file.toPath())) {
+                pathStream
+//                    .filter(path -> path.endsWith("-NOTICE.txt"))
```
no commented code please

atorok commented on buildSrc/src/main/java/org/elasticsearch/gradle/NoticeTask.java:
```diff
@@ -0,0 +1,173 @@
+                            seen.put(name, licenseFile);
+                        }
+                    });
+            } catch (IOException e) {
```
No need to catch this here.

mike2ykme replied:
These are caught as checked and re-thrown as unchecked because they are being handled inside a lambda expression inside a forEach() on the list

atorok commented on buildSrc/src/main/java/org/elasticsearch/gradle/NoticeTask.java:
```diff
@@ -0,0 +1,173 @@
+                builder.append(buffer, 0, read);
+            }
+
+        } catch (IOException e) {
```
It's generally not desirable to catch `IOException` low level, these should be bubbled up.

alpar-t reviewed (commented):
Thanks for the updates, I left some more comments.

mike2ykme committed:
Updated per: https://github.com/elastic/elasticsearch/pull/37032/files
Removed TemporaryFolder and replaced with Gradle temporaryDir
removed commented out code
replaced getText with Apache FileUtils.readFileToString()

rjernst commented on buildSrc/src/main/java/org/elasticsearch/gradle/NoticeTask.java:
```diff
@@ -0,0 +1,154 @@
+        // party component names, unaffected by the full path to the various files
+        final Map<String, File> seen = new TreeMap<>();
+
+        licensesDirs.forEach(file -> {
```
Is a forEach really necessary? A straight for loop would be just as readable, and would not require catching/rethrowing unchecked IOException

mike2ykme replied:
Oh yeah that’s a good point. I’ll make the change.

mike2ykme committed:
Updated per: https://github.com/elastic/elasticsearch/pull/37032/files
pulled loop/stream to get notices out of given directories into separate function to allow throwing of IOException

mike2ykme committed:
fixed typo in name of getFilesToAppend() method

mike2ykme commented on buildSrc/src/main/java/org/elasticsearch/gradle/NoticeTask.java:
```diff
@@ -0,0 +1,153 @@
+
+        // This is a map rather than a set so that the sort order is the 3rd
+        // party component names, unaffected by the full path to the various files
+        final Map<String, File> seen = this.getFilesToAppend(licensesDirs);
```
Replaced the stream that re-threw a checked exception with a function to make it easier to read.

mike2ykme commented on buildSrc/src/main/java/org/elasticsearch/gradle/NoticeTask.java:
```diff
@@ -0,0 +1,153 @@
+        }
+    }
+
+    private Map<String,File> getFilesToAppend(List<File> licensesDirectories) throws IOException{
```
This version will allow all exceptions to bubble up without being re-thrown.

atorok commented:
@elasticmachine test this please

atorok commented on buildSrc/src/main/java/org/elasticsearch/gradle/NoticeTask.java:
```diff
@@ -0,0 +1,153 @@
+import java.nio.file.DirectoryStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.*;
```
we don't allow * imports, this is why the build fails.

mike2ykme replied:
This has been replaced with individual imports.

atorok commented on buildSrc/src/test/java/org/elasticsearch/gradle/NoticeTaskTest.java:
```diff
@@ -0,0 +1,144 @@
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.apache.commons.io.FileUtils.write;
```
commons io is only available via a transitive dependency in `buildSrc`, it's not something we should be using, let's just stick to `Files`

mike2ykme replied:
When I tried to assemble the project with just the Files functions I ran into issues where Charsets were not matching. IE UTF-8 wouldn't always be used. I've changed the function to simply ignore bad characters using the Java IO/NIO functions.

atorok commented on buildSrc/src/main/java/org/elasticsearch/gradle/NoticeTask.java:
```diff
@@ -0,0 +1,153 @@
+            appendFileToOutput(licenseFile, name, "LICENSE", output);
+        }
+
+        Files.write(outputFile.toPath(), output.toString().getBytes());
```
should specify a char set here.

mike2ykme committed:
Removed the java.util.*; import and replaced with individual items.
Reading the files with Files.lines and such with just a Charset would fail if the file being read was not of the expected file format. I've now made it simply ignore the character to avoid any problems with the file being written back as UTF-8

mike2ykme commented on buildSrc/src/main/java/org/elasticsearch/gradle/NoticeTask.java:
```diff
@@ -0,0 +1,187 @@
+            }
+        }
+        return licensesSeen;
+    }
```
I ran into issues with using the NIO Files methods using the UTF-8 charsets failing when they were reading the notice files and would run into a "java.nio.charset.MalformedInputException: Input length = 1" when running assemble command using Gradle. So I used the decoder here to cause it to replace the bad characters with the expected bad characters for UTF-8 so it can be clearly seen what characters were replaced.

danielmitterdorfer commented:
@mike2ykme are you still intersted in getting this in?

mike2ykme commented:
Yes, is there anything I need to do for this? I thought it was just waiting on being tested.

atorok commented:
@mike2ykme I had a comment w.r.t. the bad encoding, and throwing an error and fixing them rather than ignoring it above. 
Note that you'll also have to merge master in to get the checks passing.

mike2ykme committed:
Renamed tests class to NoticeTaskTests with an (s) to ensure it won't fail cause any issues
The NoticeTask will now throw a RuntimeException indicating which file caused the problem. The MalformedInputException is wrapped inside a java.io.UncheckedException by Files.lines(),

mike2ykme commented:
Thanks! I have now removed the code that would just ignore the bad character sets. Now it will throw an exception indicating there is a problem. I have noticed that Files.lines() wraps the MalformedInputException in a java.io.UncheckedException so I have caught these and rethrown a RuntimeException with the purpose of showing which file is causing the problem to avoid having to guess at where the problem is occuring at.

atorok commented:
thanks @mike2ykme ! 
I suspect we'll now have a failing build that will highlight the bad engodings. 
We won't be able to merge this RP if it fails the build, so we'll also have to fix those files in here.

@elasticmachine test this please

javanna commented:
Do we plan on merging this or shall we close it?

atorok commented:
I'm going to close it @mike2ykme feel free to reopen if you plan to work on this again
